Fixed bug where URDF joint effort limits were not being correctly parsed

liangfok opened the pull request:
Resolves https://github.com/RobotLocomotion/drake/issues/2541

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/2544)
<!-- Reviewable:end -->

liangfok commented:
+@manuelli for feature review
+@jwnimmer-tri for platform review

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> #### Fixed bug where URDF joint effort limits were not being correctly parsed
> 
> Resolves https://github.com/RobotLocomotion/drake/issues/2541
> 
> 
> 
</details>


---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-:-KJxrOXWBDvJCHnEAdNS:-2020374368)*
<!-- Sent from Reviewable.io -->

liangfok commented:
This PR resolves a bug where the joint effort limits specified within a URDF were not being correctly parsed. The problem was the effort limits were parsed within the context of a `transmission` element whereas the effort limits were actually specified within the context of a `robot` element. To overcome this problem, I modified `DrakeJoint` to store the effort limit and updated the transmission parser to obtain the limits from `DrakeJoint`.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> +@manuelli for feature review
> +@jwnimmer-tri for platform review
</details>


---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-:-KJxr_FIH57i_-xNAdF7:-1006475358)*
<!-- Sent from Reviewable.io -->

liangfok commented:
This PR passed https://drake-jenkins.csail.mit.edu/job/experimental/2027/ and is thus ready for review.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> This PR resolves a bug where the joint effort limits specified within a URDF were not being correctly parsed. The problem was the effort limits were parsed within the context of a `transmission` element whereas the effort limits were actually specified within the context of a `robot` element. To overcome this problem, I modified `DrakeJoint` to store the effort limit and updated the transmission parser to obtain the limits from `DrakeJoint`.
</details>


---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-:-KJxxnTudFFbn776c06E:-1207194601)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
I'll take a look.

liangfok commented:
Thanks! +@amcastro-tri for feature review.

<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I'll take a look.
</details>


---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-:-KK0Au-6DQTisnb-RIXJ:619330173)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
@liangfok: thank you for identifying this problem and taking care of it!. See my comments below. My main comment is that I dont think adding "effort" to joints is the right abstraction, it doesn't belong there (notice it was not there in the first place). More on that below PTAL!

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Thanks! +@amcastro-tri for feature review.
</details>


---

Review status: 0 of 8 files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTree.cpp, line 104 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0UzExk0UKUooUgDfF:-KK0UzExk0UKUooUgDfG:-2106449770) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/RigidBodyTree.cpp#L104)):*
> ```cpp
> }
> 
> const RigidBodyActuator& RigidBodyTree::get_actuator(std::string name) const {
> ```

`const std::string&`

---

*[drake/systems/plants/RigidBodyTree.cpp, line 105 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0VcB5QtGzhVAv-zye:-KK0VcB5QtGzhVAv-zyf:-932916136) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/RigidBodyTree.cpp#L105)):*
> ```cpp
> 
> const RigidBodyActuator& RigidBodyTree::get_actuator(std::string name) const {
>   for (const auto& actuator_ptr : actuators) {
> ```

is not a pointer. Just call `actuator`?

---

*[drake/systems/plants/RigidBodyTree.cpp, line 105 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0W44KXYevchQIIIfF:-KK0W44KXYevchQIIIfG:-1806439418) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/RigidBodyTree.cpp#L105)):*
> ```cpp
> 
> const RigidBodyActuator& RigidBodyTree::get_actuator(std::string name) const {
>   for (const auto& actuator_ptr : actuators) {
> ```

Consider using `std::find`

---

*[drake/systems/plants/RigidBodyTree.h, line 691 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0WIsjhhSpzRrotN7v:-KK0WIsjhhSpzRrotN7w:1640411903) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/RigidBodyTree.h#L691)):*
> ```objc
>    *
>    * @param name The name of the rigid body actuator to get.
>    * @return A const reference to the rigid body actuator with name @p name.
> ```

consider `@returns`

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 547 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0a0_OAAkFUJO30Pzy:-KK0a0_OAAkFUJO30Pzz:-163595765) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/RigidBodyTreeURDF.cpp#L547)):*
> ```cpp
>   if (joint_body != nullptr) {
>     effort_max = joint_body->getJoint().get_joint_effort_limit_max()[0];
>     effort_min = joint_body->getJoint().get_joint_effort_limit_min()[0];
> ```

I understand what is going on here. However I don't think it is the right solution. You are using the actual Joint as a messenger to carry the effort limits in the scope of the joints description in the URDF to the scope of the transmission to instantiate a RigidBodyActuator.
What I am trying to say is, the effort limit does not belong to the Joint and therefore should not be added there even when the URDF specification sticks it in there. Reading a file format is different from the actual abstraction used in the code. Imagine supporting several file formats. Should the Joint then abide by their structure?
The effort limit belongs to the RigidBodyActuator, as it did before this PR. It is the actuator the one that cannot apply more torque beyond its capabilities. In principle (unless the materials brake) you can apply as much torque as you want to the joint, it is just a hinge. 
Would it be possible to go up one node in the URDF so that you are now at the <robot> level and from there find the <joint> with the appropriate name to extract the limits in this method (not in the `setLimits` method for the joint). 
There is a `XMLElement::Parent()` (actually inherited from `XMLNode`) that allows you to do this.

---

*[drake/systems/plants/joints/DrakeJoint.cpp, line 16 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0d3dPNG0PslNp3Cpo:-KK0d3dQuYpFF96pG7dc:-2045390393) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/joints/DrakeJoint.cpp#L16)):*
> ```cpp
>       joint_limit_max(VectorXd::Constant(
>           _num_positions, std::numeric_limits<double>::infinity())),
>       joint_effort_limit_min(VectorXd::Constant(
> ```

See above. Joint effort does not belong here but to the `RigidBodyActuator`.

---

*[drake/systems/plants/joints/DrakeJoint.cpp, line 46 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0dIYTVeY3qMasxvMh:-KK0dIYTVeY3qMasxvMi:299177559) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/joints/DrakeJoint.cpp#L46)):*
> ```cpp
> }
> 
> const Eigen::VectorXd& DrakeJoint::get_joint_effort_limit_min() const {
> ```

see above on effort not belonging to Joint.

---

*[drake/systems/plants/joints/FixedAxisOneDoFJoint.h, line 25 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0dPEc5R8OZkFsuf4F:-KK0dPEc5R8OZkFsuf4G:1341558533) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/joints/FixedAxisOneDoFJoint.h#L25)):*
> ```objc
>   double coulomb_window;
>   double effort_min;
>   double effort_max;
> ```

We shouldn't add members to our objects simply to support reading a particular file format. Notice that there is a good reason why `effort_min/max` were not members of joints in the first place. They are members of RigidBodyManipulator, that is the right abstraction.

---

*[drake/systems/plants/test/urdf_parser_test.cc, line 1 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0dldj3Y5NUuPuUNL1:-KK0dldj3Y5NUuPuUNL2:-736009945) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/test/urdf_parser_test.cc#L1)):*
> ```cpp
> #include <iostream>
> ```

+1 to the Google test!

---

*[drake/systems/plants/test/urdf_parser_test.cc, line 16 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0drI1eDtgrVGEg1cx:-KK0drI1eDtgrVGEg1cy:428383618) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/test/urdf_parser_test.cc#L16)):*
> ```cpp
>   // Defines a URDF string describing a robot with two links and one joint.
>   std::string urdf_string =
>       "<?xml version=\"1.0\" ?>"
> ```

Why a hardcoded urdf instead of using an actual file? why not to use `examples/Atlas/urdf/atlas_minimal_contact.urdf`? it has effort limits in there (so that didn't work before?). Also I like that this is one of our most complex urdf models. It'd be a nice hard test!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-:-KK0egIZTWC97Rehp2Ry:-1236622332)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 0 of 8 files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 547 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0a0_OAAkFUJO30Pzy:-KK0fW3tJiUxFE6rT34s:351487273) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/RigidBodyTreeURDF.cpp#L547)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I understand what is going on here. However I don't think it is the right solution. You are using the actual Joint as a messenger to carry the effort limits in the scope of the joints description in the URDF to the scope of the transmission to instantiate a RigidBodyActuator.
> What I am trying to say is, the effort limit does not belong to the Joint and therefore should not be added there even when the URDF specification sticks it in there. Reading a file format is different from the actual abstraction used in the code. Imagine supporting several file formats. Should the Joint then abide by their structure?
> The effort limit belongs to the RigidBodyActuator, as it did before this PR. It is the actuator the one that cannot apply more torque beyond its capabilities. In principle (unless the materials brake) you can apply as much torque as you want to the joint, it is just a hinge. 
> Would it be possible to go up one node in the URDF so that you are now at the <robot> level and from there find the <joint> with the appropriate name to extract the limits in this method (not in the `setLimits` method for the joint). 
> There is a `XMLElement::Parent()` (actually inherited from `XMLNode`) that allows you to do this.
> 
> 
</details>
Edit to the last paragraph: ...so that you are now at the level of `<robot>` and from there...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 0 of 8 files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 547 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0a0_OAAkFUJO30Pzy:-KK17DDaJk6DpiTYA5ee:-2067581089) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/RigidBodyTreeURDF.cpp#L547)):*
>

Now that you mention it, yes, I believe there is a way. I did not realize there was a `tinyxml2::XMLNode::parent()` method:

http://www.grinninglizard.com/tinyxml2docs/classtinyxml2_1_1_x_m_l_node.html#a4e39bdcf9bfafa55d04857ece6aaf64e

I will update this PR to make use of this method.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 0 of 8 files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 547 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0a0_OAAkFUJO30Pzy:-KK3bmo4q7UzffLAXwWT:1474778702) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/RigidBodyTreeURDF.cpp#L547)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> >
> 
> Now that you mention it, yes, I believe there is a way. I did not realize there was a `tinyxml2::XMLNode::parent()` method:
> 
> http://www.grinninglizard.com/tinyxml2docs/classtinyxml2_1_1_x_m_l_node.html#a4e39bdcf9bfafa55d04857ece6aaf64e
> 
> I will update this PR to make use of this method.
</details>
Yes. You don't even need to browse the doc. Everything is in `drake/thirdParty/tinyxml2/tinixml2.h`. I just found about it browsing this :-)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 0 of 8 files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 547 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0a0_OAAkFUJO30Pzy:-KK3fAR0M4prEp728kP4:235470518) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/RigidBodyTreeURDF.cpp#L547)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Yes. You don't even need to browse the doc. Everything is in `drake/thirdParty/tinyxml2/tinixml2.h`. I just found about it browsing this :-)
</details>
I forgot to mention. Once you are at the `<robot>` scope (using XMLNode::parent(), `dynamic_cast<XMLElement*>` needed), you can re-browse for joints as already done [here in `parseRobot`](https://github.com/RobotLocomotion/drake/blob/master/drake/systems/plants/RigidBodyTreeURDF.cpp#L731) and calling your `parseEfforts` instead of `parseJoint`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL.

<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> @liangfok: thank you for identifying this problem and taking care of it!. See my comments below. My main comment is that I dont think adding "effort" to joints is the right abstraction, it doesn't belong there (notice it was not there in the first place). More on that below PTAL!
</details>


---

Review status: 0 of 6 files reviewed at latest revision, 11 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTree.cpp, line 104 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0UzExk0UKUooUgDfF:-KK5ADvUtPjx5_UCIRBx:531884945) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/RigidBodyTree.cpp#L104)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> `const std::string&`
</details>
Done. I also changed the name of the method to be `GetActuatator()` since it is `O(n)` where `n` is the number of acuators in the model and is more than 5 lines long.

http://drake.mit.edu/code_style_guide.html#clarifications

---

*[drake/systems/plants/RigidBodyTree.cpp, line 105 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0VcB5QtGzhVAv-zye:-KK5AfIv4-SaYqF0D3OC:2012282870) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/RigidBodyTree.cpp#L105)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> is not a pointer. Just call `actuator`?
</details>
Done. I changed the variable to be `actuator`.

---

*[drake/systems/plants/RigidBodyTree.cpp, line 105 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0W44KXYevchQIIIfF:-KK5BGMBrSx02hO8Op3k:782259936) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/RigidBodyTree.cpp#L105)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Consider using `std::find`
</details>
Unfortunately, using `std::find()` requires overriding `operator==` in `RigidBodyActuator` to focus on `RigidBodyActuator::name`. I'm not sure that's something we want to do and thus am for now keeping the current `for` loop. If it is something we want to do, I will be happy to implement it.

---

*[drake/systems/plants/RigidBodyTree.h, line 691 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0WIsjhhSpzRrotN7v:-KK5ChZhoOpcc2KGMe0j:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/RigidBodyTree.h#L691)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> consider `@returns`
</details>
Done.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 547 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0a0_OAAkFUJO30Pzy:-KK5ClxmnUGxBX4MP47F:2083781307) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/RigidBodyTreeURDF.cpp#L547)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I forgot to mention. Once you are at the `<robot>` scope (using XMLNode::parent(), `dynamic_cast<XMLElement*>` needed), you can re-browse for joints as already done [here in `parseRobot`](https://github.com/RobotLocomotion/drake/blob/master/drake/systems/plants/RigidBodyTreeURDF.cpp#L731) and calling your `parseEfforts` instead of `parseJoint`.
</details>
Done. See the newly added method `GetActuarEffortLimit()` in RigidBodyTreeURDF.cpp.

---

*[drake/systems/plants/joints/FixedAxisOneDoFJoint.h, line 25 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0dPEc5R8OZkFsuf4F:-KK5D8UNWGwpJtTu5qdw:1740281843) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/joints/FixedAxisOneDoFJoint.h#L25)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> We shouldn't add members to our objects simply to support reading a particular file format. Notice that there is a good reason why `effort_min/max` were not members of joints in the first place. They are members of RigidBodyManipulator, that is the right abstraction.
</details>
Removed.

---

*[drake/systems/plants/test/urdf_parser_test.cc, line 16 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0drI1eDtgrVGEg1cx:-KK5DC_0KiqqZOhEaNxM:-460622735) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/test/urdf_parser_test.cc#L16)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Why a hardcoded urdf instead of using an actual file? why not to use `examples/Atlas/urdf/atlas_minimal_contact.urdf`? it has effort limits in there (so that didn't work before?). Also I like that this is one of our most complex urdf models. It'd be a nice hard test!
</details>
According to @jwnimmer-tri, we do not allow unit tests in `drake/systems/plants/test/`to reference files in `drake/examples/`.  See issue #1987 and/or pull #2533. I could make a copy of a URDF from `drake/examples/` and place it is `drake/systems/` but that also seems like bad form since keeping them in sync will be difficult.

What do you think about creating a top-level `drake/models/` directory for holding models that are shared by both stuff in `drake/examples/` and `drake/systems/`?

---

*[drake/systems/plants/joints/DrakeJoint.cpp, line 16 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0d3dPNG0PslNp3Cpo:-KK5D0nEpKtRCS2doLWI:1740281843) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/joints/DrakeJoint.cpp#L16)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> See above. Joint effort does not belong here but to the `RigidBodyActuator`.
</details>
Removed.

---

*[drake/systems/plants/joints/DrakeJoint.cpp, line 46 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0dIYTVeY3qMasxvMh:-KK5D2LVCDbESWsj-Qmk:1740281843) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/joints/DrakeJoint.cpp#L46)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> see above on effort not belonging to Joint.
</details>
Removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-:-KK5EGAYbWSj7Vo_2l6l:-84625242)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 0 of 6 files reviewed at latest revision, 11 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 481 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK5SRhna9jZx5hy6Iqb:-KK5SRhna9jZx5hy6Iqc:1901510759) ([raw file](https://github.com/robotlocomotion/drake/blob/1db8d731e4378b3e93ffeb60c2a6ff46934c624c/drake/systems/plants/RigidBodyTreeURDF.cpp#L481)):*
> ```cpp
> }
> 
> void GetActuarEffortLimit(XMLElement* node, const std::string& joint_name,
> ```

spelling error

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 0 of 6 files reviewed at latest revision, 11 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 481 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK5SRhna9jZx5hy6Iqb:-KK5rQerYrn-R-T06PjW:-795674665) ([raw file](https://github.com/robotlocomotion/drake/blob/1db8d731e4378b3e93ffeb60c2a6ff46934c624c/drake/systems/plants/RigidBodyTreeURDF.cpp#L481)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> spelling error
</details>
Fixed. Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Looks great! PTAL my last set of comments.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I addressed all reviewer comments. PTAL.
</details>


---

Review status: 0 of 6 files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTree.cpp, line 105 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0W44KXYevchQIIIfF:-KK6X282bEiCaL6iv3Dr:647328940) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/RigidBodyTree.cpp#L105)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Unfortunately, using `std::find()` requires overriding `operator==` in `RigidBodyActuator` to focus on `RigidBodyActuator::name`. I'm not sure that's something we want to do and thus am for now keeping the current `for` loop. If it is something we want to do, I will be happy to implement it.
</details>
or you could use `std::find_if` with a lambda. But I am ok with this version. I just love those one liner constructions. sorry.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 547 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0a0_OAAkFUJO30Pzy:-KK6aYvqLgN3eXjeDnhc:2111794213) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/RigidBodyTreeURDF.cpp#L547)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Done. See the newly added method `GetActuarEffortLimit()` in RigidBodyTreeURDF.cpp.
</details>
Just to be on the safe side, what about something like:
```c++
if (XMLElement* element = dynamic_cast<XMLElement*>(node->Parent())
    GetActuatorEffortLimit(element, joint_name, effort_min, effort_max);
else 
    throw std::runtime_error("Error: msg")
```

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 481 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK6eTKbUQSnNmbJj5fm:-KK6eTKbUQSnNmbJj5fn:-1191992439) ([raw file](https://github.com/robotlocomotion/drake/blob/df3a166c326921dc3a121e88bdaaece307f4f275/drake/systems/plants/RigidBodyTreeURDF.cpp#L481)):*
> ```cpp
> }
> 
> void GetActuatorEffortLimit(XMLElement* node, const std::string& joint_name,
> ```

Please add at least a brief (non-doxygen) documentation for the new method here. 

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 482 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK6dw1gLFpKqhlopolj:-KK6dw1gLFpKqhlopolk:370836303) ([raw file](https://github.com/robotlocomotion/drake/blob/df3a166c326921dc3a121e88bdaaece307f4f275/drake/systems/plants/RigidBodyTreeURDF.cpp#L482)):*
> ```cpp
> 
> void GetActuatorEffortLimit(XMLElement* node, const std::string& joint_name,
>                           double* min_effort, double* max_effort) {
> ```

I personally like to pass (non-const) references here since min/max_effort cannot be null pointers. Plus, then you don't need to dereference them when assigning a value. 

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 483 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK6dppF8sbJl0rY-Z0-:-KK6dppGDf6sK58czxBC:-279810307) ([raw file](https://github.com/robotlocomotion/drake/blob/df3a166c326921dc3a121e88bdaaece307f4f275/drake/systems/plants/RigidBodyTreeURDF.cpp#L483)):*
> ```cpp
> void GetActuatorEffortLimit(XMLElement* node, const std::string& joint_name,
>                           double* min_effort, double* max_effort) {
>   bool joint_found = false;
> ```

This flag is not needed. See below.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 519 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK6dAKhwyAdQsoXyg71:-KK6dAKhwyAdQsoXyg72:49895693) ([raw file](https://github.com/robotlocomotion/drake/blob/df3a166c326921dc3a121e88bdaaece307f4f275/drake/systems/plants/RigidBodyTreeURDF.cpp#L519)):*
> ```cpp
>         parseScalarAttribute(limit_node, "effort_min", *min_effort);
>         parseScalarAttribute(limit_node, "effort_max", *max_effort);
>       }
> ```

If you found the joint, you can `return` weather you found a `<limit>` element or not. You don't need to finish scanning siblings.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 522 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK6dWlxJe2UOKVBUUun:-KK6dWly9bNoDuiROTgm:-141090321) ([raw file](https://github.com/robotlocomotion/drake/blob/df3a166c326921dc3a121e88bdaaece307f4f275/drake/systems/plants/RigidBodyTreeURDF.cpp#L522)):*
> ```cpp
>     }
>   }
> ```

I you didn't return above when you found the joint, that means that the joint was not found. 
Then if you are at this point you can `throw`, you don't need the flag `joint_found`.

---

*[drake/systems/plants/test/urdf_parser_test.cc, line 16 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0drI1eDtgrVGEg1cx:-KK6XcE7YMKRQ79wni8B:1649785918) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/test/urdf_parser_test.cc#L16)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> According to @jwnimmer-tri, we do not allow unit tests in `drake/systems/plants/test/`to reference files in `drake/examples/`.  See issue #1987 and/or pull #2533. I could make a copy of a URDF from `drake/examples/` and place it is `drake/systems/` but that also seems like bad form since keeping them in sync will be difficult.
> 
> What do you think about creating a top-level `drake/models/` directory for holding models that are shared by both stuff in `drake/examples/` and `drake/systems/`?
</details>
I like that idea!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-:-KK6fEiTD8uhn_yUxcO8:-59605701)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 0 of 6 files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 547 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0a0_OAAkFUJO30Pzy:-KK8wCjEr9_TxcJSjnvG:1199471257) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/RigidBodyTreeURDF.cpp#L547)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Just to be on the safe side, what about something like:
> ```c++
> if (XMLElement* element = dynamic_cast<XMLElement*>(node->Parent())
>     GetActuatorEffortLimit(element, joint_name, effort_min, effort_max);
> else 
>     throw std::runtime_error("Error: msg")
> ```
</details>
I just realized it'd be more appropriate to throw a logic error like so:
```c++
throw std::logic_error("Expected a <robot> element as a parent of a <transmission> element.");
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL.

<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Looks great! PTAL my last set of comments.
</details>


---

Review status: 0 of 6 files reviewed at latest revision, 11 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTree.cpp, line 105 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0W44KXYevchQIIIfF:-KK904koGC7MvWhlrGck:285172216) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/RigidBodyTree.cpp#L105)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> or you could use `std::find_if` with a lambda. But I am ok with this version. I just love those one liner constructions. sorry.
</details>
I see. Clearly, one benefit of using that is code brevity. However, in this case, I think spelling things out using a `for` loop is more clear for those less familiar with C++. I will thus keep the current implementation. 

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 547 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0a0_OAAkFUJO30Pzy:-KK9CQdgeD0KfBAQI1XN:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/RigidBodyTreeURDF.cpp#L547)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I just realized it'd be more appropriate to throw a logic error like so:
> ```c++
> throw std::logic_error("Expected a <robot> element as a parent of a <transmission> element.");
> ```
</details>
Done.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 481 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK6eTKbUQSnNmbJj5fm:-KK97tljtV9TzAtukpyh:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/df3a166c326921dc3a121e88bdaaece307f4f275/drake/systems/plants/RigidBodyTreeURDF.cpp#L481)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Please add at least a brief (non-doxygen) documentation for the new method here. 
</details>
Done.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 482 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK6dw1gLFpKqhlopolj:-KK97pfaeCREpMCboAi7:1327039835) ([raw file](https://github.com/robotlocomotion/drake/blob/df3a166c326921dc3a121e88bdaaece307f4f275/drake/systems/plants/RigidBodyTreeURDF.cpp#L482)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I personally like to pass (non-const) references here since min/max_effort cannot be null pointers. Plus, then you don't need to dereference them when assigning a value. 
</details>
Unfortunately, the style guide prevents us from using non-const reference parameters: https://google.github.io/styleguide/cppguide.html#Reference_Arguments

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 483 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK6dppF8sbJl0rY-Z0-:-KK9Hy9ngZSaFDE-ehda:1740281843) ([raw file](https://github.com/robotlocomotion/drake/blob/df3a166c326921dc3a121e88bdaaece307f4f275/drake/systems/plants/RigidBodyTreeURDF.cpp#L483)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> This flag is not needed. See below.
</details>
Removed.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 519 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK6dAKhwyAdQsoXyg71:-KK98-hC5OCrreQGwSuH:1367517901) ([raw file](https://github.com/robotlocomotion/drake/blob/df3a166c326921dc3a121e88bdaaece307f4f275/drake/systems/plants/RigidBodyTreeURDF.cpp#L519)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> If you found the joint, you can `return` weather you found a `<limit>` element or not. You don't need to finish scanning siblings.
</details>
For some reason, I believe there's a rule where methods should only have one point of return -- at the end. Since I couldn't find such a rule in our current style guide, I changed the code as you suggest.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 522 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK6dWlxJe2UOKVBUUun:-KK99AMzoJB089LjCWVM:1287899945) ([raw file](https://github.com/robotlocomotion/drake/blob/df3a166c326921dc3a121e88bdaaece307f4f275/drake/systems/plants/RigidBodyTreeURDF.cpp#L522)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I you didn't return above when you found the joint, that means that the joint was not found. 
> Then if you are at this point you can `throw`, you don't need the flag `joint_found`.
</details>
Agreed. Fixed.

---

*[drake/systems/plants/test/urdf_parser_test.cc, line 16 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK0drI1eDtgrVGEg1cx:-KK9DVTca4oCy5aN8Bgl:1413644084) ([raw file](https://github.com/robotlocomotion/drake/blob/3ec551353f1302195e52332980ccf811c9375fed/drake/systems/plants/test/urdf_parser_test.cc#L16)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I like that idea!
</details>
OK. I'm pushing this into a future PR. Here's the issue: https://github.com/RobotLocomotion/drake/issues/2548

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-:-KK9I-2Q8SaSjKZmK_x1:-84625242)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 0 of 6 files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 482 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK6dw1gLFpKqhlopolj:-KK9eElLNF8B4WV0-9vp:-1455171165) ([raw file](https://github.com/robotlocomotion/drake/blob/df3a166c326921dc3a121e88bdaaece307f4f275/drake/systems/plants/RigidBodyTreeURDF.cpp#L482)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Unfortunately, the style guide prevents us from using non-const reference parameters: https://google.github.io/styleguide/cppguide.html#Reference_Arguments
</details>
Awesome! thank you for finding that in the style guide! TIL!

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 519 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK6dAKhwyAdQsoXyg71:-KK9eTj5Jqfp2FKq55au:-1948837079) ([raw file](https://github.com/robotlocomotion/drake/blob/df3a166c326921dc3a121e88bdaaece307f4f275/drake/systems/plants/RigidBodyTreeURDF.cpp#L519)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> For some reason, I believe there's a rule where methods should only have one point of return -- at the end. Since I couldn't find such a rule in our current style guide, I changed the code as you suggest.
</details>
I see how that could be confusing. Maybe the comment below helps to clarify?

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 522 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK6dWlxJe2UOKVBUUun:-KK9ejb5yvf1jf6YgfFk:-982858050) ([raw file](https://github.com/robotlocomotion/drake/blob/df3a166c326921dc3a121e88bdaaece307f4f275/drake/systems/plants/RigidBodyTreeURDF.cpp#L522)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Agreed. Fixed.
</details>
I would add here (and also to clarify the previous `return`) a comment like:
```c++
// If this point in the code is reached that means that the input supplied joint was not found in the urdf file. Therefore throw an exception.
```

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 482 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK9duagLdXBxXMLBjQA:-KK9duagLdXBxXMLBjQB:-882781819) ([raw file](https://github.com/robotlocomotion/drake/blob/8aacfacf08f86455878a3193999ca52f369669ac/drake/systems/plants/RigidBodyTreeURDF.cpp#L482)):*
> ```cpp
> 
> // Searches through the URDF document looking for the effort limits of a
> // particular joint. If the joint is not found, throw an exception. If the
> ```

throw-->it throws

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 484 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK9dyotzQdLlsYwxJ21:-KK9dyotzQdLlsYwxJ22:1216451934) ([raw file](https://github.com/robotlocomotion/drake/blob/8aacfacf08f86455878a3193999ca52f369669ac/drake/systems/plants/RigidBodyTreeURDF.cpp#L484)):*
> ```cpp
> // particular joint. If the joint is not found, throw an exception. If the
> // limits of the joint are not specified, do nothing. If the effort limits are
> // specified, save the effort limits in \p min_effort and \p max_effort.
> ```

Notice this is not doxygen (public API) documentation. So really you don't need to use doxygen annotations. However, I wouldn't know how to document parameters then. Is there a best practice?

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 539 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK9fMMki3D_nSeV_VPe:-KK9fMMki3D_nSeV_VPf:-582107862) ([raw file](https://github.com/robotlocomotion/drake/blob/8aacfacf08f86455878a3193999ca52f369669ac/drake/systems/plants/RigidBodyTreeURDF.cpp#L539)):*
> ```cpp
>   }
> 
>   throw std::logic_error(
> ```

`std::runtime_error`? if the joint is not found then there is something wrong with the urdf file (unless of course we have a bug somewhere).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I addressed all reviewer comments. PTAL.
</details>


---

Review status: 0 of 6 files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 482 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK6dw1gLFpKqhlopolj:-KK9ldFXeceOELWZbKNQ:1328085673) ([raw file](https://github.com/robotlocomotion/drake/blob/df3a166c326921dc3a121e88bdaaece307f4f275/drake/systems/plants/RigidBodyTreeURDF.cpp#L482)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Awesome! thank you for finding that in the style guide! TIL!
</details>
No problem.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 519 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK6dAKhwyAdQsoXyg71:-KK9kj2MWX1lBb9L7r6F:826581716) ([raw file](https://github.com/robotlocomotion/drake/blob/df3a166c326921dc3a121e88bdaaece307f4f275/drake/systems/plants/RigidBodyTreeURDF.cpp#L519)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I see how that could be confusing. Maybe the comment below helps to clarify?
</details>
Done. Yes, I think it's more clear now.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 522 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK6dWlxJe2UOKVBUUun:-KK9kliS1T_HhvcTJHjM:1508991532) ([raw file](https://github.com/robotlocomotion/drake/blob/df3a166c326921dc3a121e88bdaaece307f4f275/drake/systems/plants/RigidBodyTreeURDF.cpp#L522)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I would add here (and also to clarify the previous `return`) a comment like:
> ```c++
> // If this point in the code is reached that means that the input supplied joint was not found in the urdf file. Therefore throw an exception.
> ```
</details>
Done. Thanks.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 482 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK9duagLdXBxXMLBjQA:-KK9lR9ZTX9MPppossbd:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/8aacfacf08f86455878a3193999ca52f369669ac/drake/systems/plants/RigidBodyTreeURDF.cpp#L482)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> throw-->it throws
</details>
Done.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 484 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK9dyotzQdLlsYwxJ21:-KK9j2CRqAGUdNDRP7Li:1076162698) ([raw file](https://github.com/robotlocomotion/drake/blob/8aacfacf08f86455878a3193999ca52f369669ac/drake/systems/plants/RigidBodyTreeURDF.cpp#L484)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Notice this is not doxygen (public API) documentation. So really you don't need to use doxygen annotations. However, I wouldn't know how to document parameters then. Is there a best practice?
</details>
Not yet. At least not that I'm aware of. I updated https://github.com/RobotLocomotion/drake/issues/2051 to include this desire.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 539 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK9fMMki3D_nSeV_VPe:-KK9kyEVp9nxzrZ4Sm8w:-933856791) ([raw file](https://github.com/robotlocomotion/drake/blob/8aacfacf08f86455878a3193999ca52f369669ac/drake/systems/plants/RigidBodyTreeURDF.cpp#L539)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> `std::runtime_error`? if the joint is not found then there is something wrong with the urdf file (unless of course we have a bug somewhere).
</details>
Since getting to this point in the code indicates a problem with the URDF file, I think `logic_error` is correct -- it's a logical error within the URDF file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-:-KK9lXvMsMhhKsX2w5Ym:-84625242)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 539 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK9fMMki3D_nSeV_VPe:-KKAAChMIESA4NkrAB0m:696678921) ([raw file](https://github.com/robotlocomotion/drake/blob/8aacfacf08f86455878a3193999ca52f369669ac/drake/systems/plants/RigidBodyTreeURDF.cpp#L539)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Since getting to this point in the code indicates a problem with the URDF file, I think `logic_error` is correct -- it's a logical error within the URDF file.
</details>
`std::logic_error` are to signal problems with the logic [within our code](http://en.cppreference.com/w/cpp/error/logic_error), not with "logic" errors made by the user (who wrote the urdf file).  

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 540 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KKA9szgTg2hPvbmHd6i:-KKA9szgTg2hPvbmHd6j:1931400182) ([raw file](https://github.com/robotlocomotion/drake/blob/3895ac3f0d19f847ec401828f497014ecb4fd12a/drake/systems/plants/RigidBodyTreeURDF.cpp#L540)):*
> ```cpp
> 
>   // If this point in the code is reached, that means no joint named
>   // joint_name was not found in the URDF. Therefore throw an exception.
> ```

double negation. Remove "not"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
two small comments to be addressed and then it <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I addressed all reviewer comments. PTAL.
</details>


---

Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-:-KKAB8GXsdYckCdzW5CA:1094742152)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed your comment but have one question (see below).

<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> two small comments to be addressed and then it :lgtm_strong: 
</details>


---

Review status: 0 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 539 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK9fMMki3D_nSeV_VPe:-KKAD_nnY5NYJjbYiaaS:2064691199) ([raw file](https://github.com/robotlocomotion/drake/blob/8aacfacf08f86455878a3193999ca52f369669ac/drake/systems/plants/RigidBodyTreeURDF.cpp#L539)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> `std::logic_error` are to signal problems with the logic [within our code](http://en.cppreference.com/w/cpp/error/logic_error), not with "logic" errors made by the user (who wrote the urdf file).  
</details>
Fixed. Thanks. Should the above exception that's thrown when the joint tag is missing a name attribute also be a `runtime_error`?

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 540 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KKA9szgTg2hPvbmHd6i:-KKADI4GBi_RQ3_jxAlc:-795674665) ([raw file](https://github.com/robotlocomotion/drake/blob/3895ac3f0d19f847ec401828f497014ecb4fd12a/drake/systems/plants/RigidBodyTreeURDF.cpp#L540)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> double negation. Remove "not"
</details>
Fixed. Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-:-KKADs7jXWr9TOiw1nsF:-1509631667)*
<!-- Sent from Reviewable.io -->

liangfok commented:
This PR passed https://drake-jenkins.csail.mit.edu/job/experimental/2046/ and is thus ready for final review / merge.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I addressed your comment but have one question (see below).
</details>


---

Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-:-KKAb5k9fXsAl4J-xmEE:1355314825)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Since it's a Monday, -@jwnimmer-tri and +@sherm1 for platform review.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> This PR passed https://drake-jenkins.csail.mit.edu/job/experimental/2046/ and is thus ready for final review / merge.
</details>


---

Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-:-KKAgJxwfEUBrQl_MsOz:-1368141821)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Platform review: looks good -- a few minor comments.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Since it's a Monday, -@jwnimmer-tri and +@sherm1 for platform review.
</details>


---

Reviewed 4 of 7 files at r4, 1 of 2 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 506 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KKB4yTQaUACpI85ktZ6:-KKB4yTQaUACpI85ktZ7:1064764248) ([raw file](https://github.com/robotlocomotion/drake/blob/b4b5664d0e1a121fad6f665598b647e6447e83b6/drake/systems/plants/RigidBodyTreeURDF.cpp#L506)):*
> ```cpp
>     {
>       const char* attr = joint_node->Attribute("drake_ignore");
>       if (attr && strcmp(attr, "true") == 0) {
> ```

Parenthesize to make precedence clear. `attr && (std::strcmp(attr, "true")==0)`. Should use `std::` here unless there is a `using` above I didn't see.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 543 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KKB5xLnSqnbUjN4gvWO:-KKB5xLnSqnbUjN4gvWP:678074825) ([raw file](https://github.com/robotlocomotion/drake/blob/b4b5664d0e1a121fad6f665598b647e6447e83b6/drake/systems/plants/RigidBodyTreeURDF.cpp#L543)):*
> ```cpp
>   // joint_name was found in the URDF. Therefore throw an exception.
>   throw std::runtime_error(
>       "GetActuarEffortLimit: ERROR: Unable to find joint \"" + joint_name +
> ```

typo

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 562 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KKB63AM97u_3fVfIxb7:-KKB63AM97u_3fVfIxb8:-568293819) ([raw file](https://github.com/robotlocomotion/drake/blob/b4b5664d0e1a121fad6f665598b647e6447e83b6/drake/systems/plants/RigidBodyTreeURDF.cpp#L562)):*
> ```cpp
>           "RigidBodyTreeURDF.cpp: parseTransmission: ERROR: Transmission "
>           "element "
>           "is missing the type child.");
> ```

Combine with previous line.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-:-KKB7CAbb6rSgrjt8_Ez:-543010663)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL.

<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Platform review: looks good -- a few minor comments.
</details>


---

Review status: 5 of 6 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 506 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KKB4yTQaUACpI85ktZ6:-KKBMvuZIgOyaM68683i:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/b4b5664d0e1a121fad6f665598b647e6447e83b6/drake/systems/plants/RigidBodyTreeURDF.cpp#L506)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Parenthesize to make precedence clear. `attr && (std::strcmp(attr, "true")==0)`. Should use `std::` here unless there is a `using` above I didn't see.
</details>
Done.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 543 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KKB5xLnSqnbUjN4gvWO:-KKBMcqY0aD9ovARD63w:-795674652) ([raw file](https://github.com/robotlocomotion/drake/blob/b4b5664d0e1a121fad6f665598b647e6447e83b6/drake/systems/plants/RigidBodyTreeURDF.cpp#L543)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> typo
</details>
Fixed. Thanks.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 562 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KKB63AM97u_3fVfIxb7:-KKBMhKpA5wU7FZkw__R:1758359711) ([raw file](https://github.com/robotlocomotion/drake/blob/b4b5664d0e1a121fad6f665598b647e6447e83b6/drake/systems/plants/RigidBodyTreeURDF.cpp#L562)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Combine with previous line.
</details>
Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-:-KKBNejpCIjpf3ITXCXt:-84625242)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 5 of 6 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 539 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK9fMMki3D_nSeV_VPe:-KKC0bJhGi-g4cQknm3S:975289285) ([raw file](https://github.com/robotlocomotion/drake/blob/8aacfacf08f86455878a3193999ca52f369669ac/drake/systems/plants/RigidBodyTreeURDF.cpp#L539)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Fixed. Thanks. Should the above exception that's thrown when the joint tag is missing a name attribute also be a `runtime_error`?
</details>
I would say yes, a runtime_error, good catch. Since in that scope you are in a `<joint>` and the fact that a `name` is missing would indicate a user error forgetting to specify it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 5 of 6 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 539 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KK9fMMki3D_nSeV_VPe:-KKC6eBIT-MpWQ40XZjl:242947392) ([raw file](https://github.com/robotlocomotion/drake/blob/8aacfacf08f86455878a3193999ca52f369669ac/drake/systems/plants/RigidBodyTreeURDF.cpp#L539)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I would say yes, a runtime_error, good catch. Since in that scope you are in a `<joint>` and the fact that a `name` is missing would indicate a user error forgetting to specify it.
</details>
Done. I changed the exception to be a `std::runtime_exception`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544)*
<!-- Sent from Reviewable.io -->

liangfok commented:
This PR passed https://drake-jenkins.csail.mit.edu/job/experimental/2066/ and is thus ready for review.

---

Review status: 5 of 6 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-:-KKCUVGr38G2S2WNs93p:658136124)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Excelent! <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Review status: 5 of 6 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-:-KKDa6Y0OqySvh3WWTQ7:1006444422)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Down to one typo ...

---

Reviewed 1 of 1 files at r12.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 543 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KKB5xLnSqnbUjN4gvWO:-KKEwIYlBruw_rcjgx69:606530449) ([raw file](https://github.com/robotlocomotion/drake/blob/b4b5664d0e1a121fad6f665598b647e6447e83b6/drake/systems/plants/RigidBodyTreeURDF.cpp#L543)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Fixed. Thanks.
</details>
Still looks the same? I meant GetActuarEffort -> GetActuatorEffort.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-:-KKEwVyAIZPwAZlu897R:1123410616)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I fixed the typo. PTAL. Thanks!

---

Review status: 5 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 543 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-KKB5xLnSqnbUjN4gvWO:-KKF5RIhY-DiyQWO2GzR:-1828919090) ([raw file](https://github.com/robotlocomotion/drake/blob/b4b5664d0e1a121fad6f665598b647e6447e83b6/drake/systems/plants/RigidBodyTreeURDF.cpp#L543)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Still looks the same? I meant GetActuarEffort -> GetActuatorEffort.
</details>
Oops! Fixed in next revision.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-:-KKF5TlpaNXPcBWHkCRj:2094940618)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r13.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2544#-:-KKFO0YDyIRa8Q2cBBmT:1417088721)*
<!-- Sent from Reviewable.io -->
